File: mill-example/oystercard/src/oystercard/java/Amount.java
package oystercard.java;

import java.util.Locale;

public final class Amount {

    public final int valueInPence;

    public int getValueInPence() {
        return valueInPence;
    }
    public static Amount Zero = inPence(0);

    private Amount(int pence) {
        valueInPence = pence;
    }

    public static Amount inPence(int pence) {
        return new Amount(pence);
    }

    public static Amount inPounds(int pounds) {
        return inPence(pounds * 100);
    }

    public Amount plus(Amount other) {
        return new Amount(valueInPence + other.valueInPence);
    }

    public Amount minus(Amount other) {
        return new Amount(valueInPence - other.valueInPence);
    }

    @Override
    public boolean equals(Object other) {
        if (other instanceof Amount) {
            return ((Amount) other).valueInPence == valueInPence;
        } else {
            return false;
        }
    }

    @Override
    public int hashCode() {
        return valueInPence;
    }

    @Override
    public String toString() {
        return show(Locale.getDefault());
    }

    public String show(Locale locale) {
        return java.text.NumberFormat.getCurrencyInstance(locale).format(valueInPence / 100.0);
    }
}

File: mill-example/oystercard/src/oystercard/scala/Amount.scala
package oystercard.scala

import java.util.Locale

class Amount private(val valueInPence: Int) {
  def -(other: Amount): Amount = Amount.inPence(valueInPence - other.valueInPence)

  def +(other: Amount): Amount = Amount.inPence(valueInPence + other.valueInPence)

  override def toString: String = show(Locale.getDefault)

  def show(locale: Locale): String = Amount.fmt(valueInPence, locale)

  override def hashCode: Zone = valueInPence

  override def equals(other: Any) = other match {
    case Amount(pence) => valueInPence == pence
    case _ => false
  }
}

object Amount {
  val Zero = inPence(0)

  def unapply(amount: Amount) = Option(amount.valueInPence)

  def inPence(pence: Int): Amount = new Amount(pence)

  def inPounds(pounds: Int): Amount = inPence(pounds * 100)

  def fmt(valueInPence: Int, locale: Locale): String = java.text.NumberFormat.getCurrencyInstance(locale).format(valueInPence / 100.0)
}